### PR TITLE
finish edit,show profile

### PR DESCRIPTION
**Pull request by phudinhdanh**

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -26,6 +26,20 @@ def create
     end
   end
 
+def edit
+    @user = User.find(params[:id])
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: ()

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -26,6 +26,20 @@ def create
...
+  end
+
+  def update
+    @user = User.find(params[:id])
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: ()

File: app/models/user.rb [hunk 3]
```diff
@@ -8,7 +8,9 @@ class User < ActiveRecord::Base
                     uniqueness: { case_sensitive: false }
   has_secure_password
   validates :password, length: { minimum: 6 }, on: [:create]
-
+  has_attached_file :avatar, :styles => { :medium => "300x300>",
```
Comment (chuanhtuan) [inline app/models/user.rb]: =>

File: app/views/users/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,29 @@
+<% provide(:title, "Edit user") %>
+<section>
+<h1>Update your profile</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user) do |f| %>
```
Comment (chuanhtuan) [inline app/views/users/edit.html.erb]: ()

File: app/views/users/edit.html.erb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    </div>
+  -->
+    <div class="form-group">
+        <%= f.label :avatar, class: 'col-sm-2 control-label'  %>
+        <div class="col-sm-6">
+            <%= f.file_field :avatar %>
+        </div>
+    </div>
+
+      <%= f.submit "Save changes", class: "btn btn-primary" %>
```
Comment (chuanhtuan) [inline app/views/users/edit.html.erb]: i18n

File: app/views/users/show.html.erb [hunk 6]
```diff
@@ -3,9 +3,33 @@
 	<div class="row">
 	  <aside class="col-md-4">
 	    <section class="user_info">
-	      <h1>
-	        <%= @user.name %>
-	      </h1>
+	      <table style="width:100%">
+	        <%= image_tag @user.avatar %>
+	          <tr>
+			    <td>Name</td>
```
Comment (chuanhtuan) [inline app/views/users/show.html.erb]: i18n

File: app/models/user.rb [hunk 7]
```diff
@@ -8,7 +8,9 @@ class User < ActiveRecord::Base
                     uniqueness: { case_sensitive: false }
   has_secure_password
   validates :password, length: { minimum: 6 }, on: [:create]
-
+  has_attached_file :avatar, :styles => { medium:  "300x300>",
```
Comment (chuanhtuan) [inline app/models/user.rb]: em không được dùng =>
xem lại coding convention

File: app/views/sessions/new.html.erb [hunk 8]
```diff
@@ -20,7 +20,7 @@
         <%= f.submit I18n.t 'log_in.value', class: "btn btn-primary" %>
       <% end %>
 
-      <p>New user? <%= link_to I18n.t 'sign_up.value', signup_path %></p>
+      <p>New user? <%= link_to "sign up now", signup_path %></p>
```
Comment (chuanhtuan) [inline app/views/sessions/new.html.erb]: i18n

File: Gemfile [hunk 9]
```diff
@@ -31,7 +31,8 @@ gem 'bootstrap-will_paginate'
 
 # Use ActiveModel has_secure_password
 gem 'bcrypt', '~> 3.1.7'
-
+# Use paperclip
```
Comment (dieunb) [inline Gemfile]: remove all remark

File: Gemfile [hunk 10]
```diff
@@ -31,7 +31,8 @@ gem 'bootstrap-will_paginate'
...
+gem 'paperclip'
```
Comment (dieunb) [inline Gemfile]: double quote pls

File: app/controllers/application_controller.rb [hunk 11]
```diff
@@ -2,5 +2,12 @@ class ApplicationController < ActionController::Base
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+
+	protected
```
Comment (dieunb) [inline app/controllers/application_controller.rb]: indent, should 2 spaces

File: app/controllers/application_controller.rb [hunk 12]
```diff
@@ -2,5 +2,12 @@ class ApplicationController < ActionController::Base
...
+
+	def configure_permitted_parameters
+	#     devise_parameter_sanitizer.for(:sign_up) { |u| u.permit(:name, :email, :password) }
```
Comment (dieunb) [inline app/controllers/application_controller.rb]: remove all remark

File: app/controllers/application_controller.rb [hunk 13]
```diff
@@ -2,5 +2,12 @@ class ApplicationController < ActionController::Base
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+
+	protected
+
+	def configure_permitted_parameters
```
Comment (dieunb) [inline app/controllers/application_controller.rb]: indent

File: app/controllers/application_controller.rb [hunk 14]
```diff
@@ -2,5 +2,12 @@ class ApplicationController < ActionController::Base
...
+	#     devise_parameter_sanitizer.for(:sign_up) { |u| u.permit(:name, :email, :password) }
+	    devise_parameter_sanitizer.for(:account_update) { |u| u.permit(:name, :avatar) }
```
Comment (dieunb) [inline app/controllers/application_controller.rb]: * indent
* convention, should `{|u| u.permit :name, :avatar}`

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -26,6 +26,20 @@ def create
     end
   end
 
+def edit
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: indent

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -26,6 +26,20 @@ def create
...
+  end
+
+  def update
+    @user = User.find params[:id]
+
+    if @user.update_attributes! name: params[:user][:name]
+      flash[:success] = I18n.t 'profile.updated'
+      redirect_to @user
+    else
+      render 'edit'
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: double quote

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -26,6 +26,20 @@ def create
...
 
+def edit
+    @user = User.find params[:id] 
+  end
+
+  def update
+    @user = User.find params[:id]
+
+    if @user.update_attributes! name: params[:user][:name]
+      flash[:success] = I18n.t 'profile.updated'
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: double quote

File: app/helpers/users_helper.rb [hunk 18]
```diff
@@ -0,0 +1,10 @@
+module UsersHelper
+
+  # Returns the Gravatar for the given user.
```
Comment (dieunb) [inline app/helpers/users_helper.rb]: remove all remark

File: app/helpers/users_helper.rb [hunk 19]
```diff
@@ -0,0 +1,10 @@
...
+  def gravatar_for(user, options = { size: 80 })
```
Comment (dieunb) [inline app/helpers/users_helper.rb]: * remove `()`
* Convention, should `{size: 80}`

File: app/helpers/users_helper.rb [hunk 20]
```diff
@@ -0,0 +1,10 @@
...
+    gravatar_id = Digest::MD5::hexdigest(user.email.downcase)
+    size = options[:size]
+    gravatar_url = "https://secure.gravatar.com/avatar/#{gravatar_id}?s=#{size}"
+    image_tag(gravatar_url, alt: user.name, class: "gravatar")
```
Comment (dieunb) [inline app/helpers/users_helper.rb]: remove `()`

File: app/models/user.rb [hunk 21]
```diff
@@ -8,7 +8,9 @@ class User < ActiveRecord::Base
                     uniqueness: { case_sensitive: false }
   has_secure_password
   validates :password, length: { minimum: 6 }, on: [:create]
-
+  has_attached_file :avatar, styles: { medium:  "300x300>",
```
Comment (dieunb) [inline app/models/user.rb]: convention 
```
+  has_attached_file :avatar, styles: {medium:  "300x300>", 
+                     thumb:  "100x100#"}, default_url: "/images/:style/missing.png"
```

File: app/models/user.rb [hunk 22]
```diff
@@ -8,7 +8,9 @@ class User < ActiveRecord::Base
                     uniqueness: { case_sensitive: false }
   has_secure_password
   validates :password, length: { minimum: 6 }, on: [:create]
-
+  has_attached_file :avatar, styles: { medium:  "300x300>", 
+                     thumb:  "100x100#" }, :default_url => "/images/:style/missing.png"
+  validates_attachment_content_type :avatar, :content_type => /\Aimage\/.*\Z/
```
Comment (dieunb) [inline app/models/user.rb]: convention

File: app/models/user.rb [hunk 23]
```diff
@@ -8,7 +8,9 @@ class User < ActiveRecord::Base
...
   # Returns the hash digest of the given string.
```
Comment (dieunb) [inline app/models/user.rb]: remove all remark

File: app/views/sessions/new.html.erb [hunk 24]
```diff
@@ -17,10 +17,10 @@
           <span>Remember me on this computer</span>
         <% end %>
 
-        <%= f.submit I18n.t 'log_in.value', class: "btn btn-primary" %>
+        <%= f.submit t('log_in.value'), class: "btn btn-primary" %>
```
Comment (dieunb) [inline app/views/sessions/new.html.erb]: double quote, apply it on for all your code. Remember it.

File: app/views/sessions/new.html.erb [hunk 25]
```diff
@@ -17,10 +17,10 @@
...
       <% end %>
 
-      <p>New user? <%= link_to I18n.t 'sign_up.value', signup_path %></p>
+      <p>New user? <%= link_to t('sign_up.value') , signup_path %></p>
```
Comment (dieunb) [inline app/views/sessions/new.html.erb]: double quote

File: app/views/users/edit.html.erb [hunk 26]
```diff
@@ -0,0 +1,29 @@
+<% provide(:title, "Edit user") %>
```
Comment (dieunb) [inline app/views/users/edit.html.erb]: remove `()`

File: app/views/users/edit.html.erb [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+<section>
+<h1>Update your profile</h1>
```
Comment (dieunb) [inline app/views/users/edit.html.erb]: i18n

File: app/views/users/edit.html.erb [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user do |f| %>
+      <%= render 'shared/error_messages' %>
+
+      <%= f.label :name %>
+      <%= f.text_field :name, class: 'form-control' %>
+
+ <!--    <div class="gravatar_edit">
```
Comment (dieunb) [inline app/views/users/edit.html.erb]: remove if you don't want use it.

File: app/views/users/edit.html.erb [hunk 29]
```diff
@@ -0,0 +1,29 @@
+<% provide(:title, "Edit user") %>
+<section>
+<h1>Update your profile</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user do |f| %>
+      <%= render 'shared/error_messages' %>
```
Comment (dieunb) [inline app/views/users/edit.html.erb]: double quote

File: app/views/users/edit.html.erb [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+
+      <%= f.label :name %>
+      <%= f.text_field :name, class: 'form-control' %>
```
Comment (dieunb) [inline app/views/users/edit.html.erb]: double quote

File: app/views/users/edit.html.erb [hunk 31]
```diff
@@ -0,0 +1,29 @@
...
+
+ <!--    <div class="gravatar_edit">
+      <%= gravatar_for @user %>
+      <a href="http://gravatar.com/emails">change</a>
+    </div>
+  -->
+    <div class="form-group">
+        <%= f.label :avatar, class: 'col-sm-2 control-label'  %>
```
Comment (dieunb) [inline app/views/users/edit.html.erb]: double quote

File: app/views/users/edit.html.erb [hunk 32]
```diff
@@ -0,0 +1,29 @@
...
+        <div class="col-sm-6">
+            <%= f.file_field :avatar %>
```
Comment (dieunb) [inline app/views/users/edit.html.erb]: indent

File: app/views/users/edit.html.erb [hunk 33]
```diff
@@ -0,0 +1,29 @@
...
+        </div>
+    </div>
+
+      <%= f.submit I18n.t 'submit.save', class: "btn btn-primary" %>
```
Comment (dieunb) [inline app/views/users/edit.html.erb]: double quote

File: app/views/users/show.html.erb [hunk 34]
```diff
@@ -3,9 +3,33 @@
 	<div class="row">
 	  <aside class="col-md-4">
 	    <section class="user_info">
-	      <h1>
-	        <%= @user.name %>
-	      </h1>
+	      <table style="width:100%">
+	        <%= image_tag @user.avatar %>
+	          <tr>
+			    <td><%=I18n.t 'edit.name' %></td>
```
Comment (dieunb) [inline app/views/users/show.html.erb]: * double quote
* Indent

File: app/views/users/show.html.erb [hunk 35]
```diff
@@ -3,9 +3,33 @@
...
+			    <td><%= @user.name %></td> 
+			  </tr>
+			  <tr>
+			    <td><%=I18n.t 'edit.email'%></td>
```
Comment (dieunb) [inline app/views/users/show.html.erb]: double quote

File: app/views/users/show.html.erb [hunk 36]
```diff
@@ -3,9 +3,33 @@
...
+			    <td><%= @user.email %></td> 
+			  </tr>
+			  <tr>
+			    <td><%=I18n.t 'edit.following'%></td>
```
Comment (dieunb) [inline app/views/users/show.html.erb]: double quote

File: app/views/users/show.html.erb [hunk 37]
```diff
@@ -3,9 +3,33 @@
...
+			    <td><%=I18n.t 'edit.dratf1'%></td>
```
Comment (dieunb) [inline app/views/users/show.html.erb]: double quote

File: app/views/users/show.html.erb [hunk 38]
```diff
@@ -3,9 +3,33 @@
...
+			  <tr>
+			    <td><%=I18n.t 'edit.email'%></td>
+			    <td><%= @user.email %></td> 
+			  </tr>
+			  <tr>
+			    <td><%=I18n.t 'edit.following'%></td>
+			    <td><%=I18n.t 'edit.dratf1'%></td> 
+			  </tr>
+			  <tr>
+			    <td><%=I18n.t 'edit.followers'%></td>
```
Comment (dieunb) [inline app/views/users/show.html.erb]: double quote

File: app/views/users/show.html.erb [hunk 39]
```diff
@@ -3,9 +3,33 @@
...
+			    <td><%=I18n.t 'edit.dratf2'%></td>
```
Comment (dieunb) [inline app/views/users/show.html.erb]: double quote

File: app/views/users/show.html.erb [hunk 40]
```diff
@@ -3,9 +3,33 @@
...
+			  <tr>
+			    <td><%=I18n.t 'edit.following'%></td>
+			    <td><%=I18n.t 'edit.dratf1'%></td> 
+			  </tr>
+			  <tr>
+			    <td><%=I18n.t 'edit.followers'%></td>
+			    <td><%=I18n.t 'edit.dratf2'%></td> 
+			  </tr>
+			  <tr>
+			    <td><%=I18n.t 'edit.learned'%></td>
```
Comment (dieunb) [inline app/views/users/show.html.erb]: double quote

File: app/views/users/show.html.erb [hunk 41]
```diff
@@ -3,9 +3,33 @@
...
+			    <td><%=I18n.t 'edit.dratf3'%></td>
```
Comment (dieunb) [inline app/views/users/show.html.erb]: double quote

File: app/views/users/show.html.erb [hunk 42]
```diff
@@ -3,9 +3,33 @@
...
+			  <tr>
+			    <td><%=I18n.t 'edit.followers'%></td>
+			    <td><%=I18n.t 'edit.dratf2'%></td> 
+			  </tr>
+			  <tr>
+			    <td><%=I18n.t 'edit.learned'%></td>
+			    <td><%=I18n.t 'edit.dratf3'%></td> 
+			  </tr>
+			  <tr>
+			    <td><%=I18n.t 'edit.recently'%></td>
```
Comment (dieunb) [inline app/views/users/show.html.erb]: double quote

File: app/views/users/show.html.erb [hunk 43]
```diff
@@ -3,9 +3,33 @@
...
+			    <td><%=I18n.t 'edit.dratf4'%></td>
```
Comment (dieunb) [inline app/views/users/show.html.erb]: double quote

File: app/controllers/application_controller.rb [hunk 44]
```diff
@@ -2,5 +2,9 @@ class ApplicationController < ActionController::Base
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+	protected
```
Comment (dieunb) [inline app/controllers/application_controller.rb]: indent

File: app/controllers/application_controller.rb [hunk 45]
```diff
@@ -2,5 +2,9 @@ class ApplicationController < ActionController::Base
...
+	def configure_permitted_parameters
+    devise_parameter_sanitizer.for(:account_update) { |u| u.permit :name, :avatar }
```
Comment (dieunb) [inline app/controllers/application_controller.rb]: convention, should change `{|u| u.permit :name, :avatar}`

File: app/helpers/users_helper.rb [hunk 46]
```diff
@@ -0,0 +1,8 @@
+module UsersHelper
+  def gravatar_for user, options = {size: 80 }
```
Comment (dieunb) [inline app/helpers/users_helper.rb]: `{size: 80}`

File: app/models/user.rb [hunk 47]
```diff
@@ -1,42 +1,41 @@
...
+  validates :email, presence: true, length: {maximum: 25},
+                    format: {with: VALID_EMAIL_REGEX},
+                    uniqueness: {case_sensitive: false}
   has_secure_password
-  validates :password, length: { minimum: 6 }, on: [:create]
-
-  # Returns the hash digest of the given string.
+  validates :password, length: {minimum: 6}, on: [:create]
+  has_attached_file :avatar, styles: {medium: "300x300>", thumb: "100x100#"}, 
+                    :default_url => "/images/:style/missing.png"
```
Comment (dieunb) [inline app/models/user.rb]: `default_url: "/images/:style/missing.png"`

File: app/views/sessions/new.html.erb [hunk 48]
```diff
@@ -17,10 +17,10 @@
           <span>Remember me on this computer</span>
         <% end %>
 
-        <%= f.submit I18n.t 'log_in.value', class: "btn btn-primary" %>
+        <%= f.submit t("log_in.value"), class: "btn btn-primary" %>
       <% end %>
 
-      <p>New user? <%= link_to I18n.t 'sign_up.value', signup_path %></p>
+      <p>New user? <%= link_to t("sign_up.value") , signup_path %></p>
```
Comment (dieunb) [inline app/views/sessions/new.html.erb]: i18n

File: app/views/users/edit.html.erb [hunk 49]
```diff
@@ -0,0 +1,24 @@
...
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user do |f| %>
+      <%= render "shared/error_messages" %>
+
+      <%= f.label :name %>
+      <%= f.text_field :name, class: "form-control" %>
+
+    <div class="form-group">
```
Comment (dieunb) [inline app/views/users/edit.html.erb]: indent

File: app/views/users/edit.html.erb [hunk 50]
```diff
@@ -0,0 +1,24 @@
...
+        <%= f.label :avatar, class: "col-sm-2 control-label"  %>
+        <div class="col-sm-6">
+          <%= f.file_field :avatar %>
+        </div>
+    </div>
```
Comment (dieunb) [inline app/views/users/edit.html.erb]: indent

File: app/views/users/show.html.erb [hunk 51]
```diff
@@ -1,12 +1,36 @@
...
-	    </section>
-	  </aside>
-	</div>
+<div class="row">
+  <aside class="col-md-4">
+    <section class="user_info">
+      <table style="width:100%">
+        <%= image_tag @user.avatar %>
+	      <tr>
+		    <td><%=I18n.t "edit.name" %></td>
```
Comment (dieunb) [inline app/views/users/show.html.erb]: indent

File: app/controllers/categories_controller.rb [hunk 52]
```diff
@@ -0,0 +1,8 @@
+class CategoriesController < ApplicationController
+  def index
+    @categories = Categories.paginate page: params[:page]
+  end
+  def show
+    @categories = Categories.find(params[:id])
```
Comment (dieunb) [inline app/controllers/categories_controller.rb]: remove `()`

File: app/controllers/categories_controller.rb [hunk 53]
```diff
@@ -0,0 +1,8 @@
+class CategoriesController < ApplicationController
+  def index
+    @categories = Categories.paginate page: params[:page]
+  end
+  def show
```
Comment (dieunb) [inline app/controllers/categories_controller.rb]: add new line above this action

File: app/views/users/edit.html.erb [hunk 54]
```diff
@@ -0,0 +1,24 @@
...
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user do |f| %>
+      <%= render "shared/error_messages" %>
+
+      <%= f.label :name %>
+      <%= f.text_field :name, class: "form-control" %>
+
+  <div class="form-group">
```
Comment (dieunb) [inline app/views/users/edit.html.erb]: indent

File: app/views/users/edit.html.erb [hunk 55]
```diff
@@ -0,0 +1,24 @@
...
+      <%= f.label :avatar, class: "col-sm-2 control-label"  %>
+      <div class="col-sm-6">
+        <%= f.file_field :avatar %>
+      </div>
+  </div>
```
Comment (dieunb) [inline app/views/users/edit.html.erb]: indent

File: app/views/users/edit.html.erb [hunk 56]
```diff
@@ -0,0 +1,24 @@
...
+
+  <%= f.submit I18n.t "submit.save", class: "btn btn-primary" %>
```
Comment (dieunb) [inline app/views/users/edit.html.erb]: indent

**Pull request merged**